client/components/site-verticals-suggestion-search/index.jsx

```diff
@@ -52,9 +50,61 @@ export class SiteVerticalsSuggestionSearch extends Component {
 		this.state = {
 			railcar: this.getNewRailcar(),
 			candidateVerticals: [],
+			isSuggestionSelected: false,
+			searchValue: props.searchValue,
 		};
+		this.updateVerticalDataDebounced = props.bindDebounce( this.updateVerticalData );
 	}
 
+	componentDidUpdate( prevProps ) {
+		// The suggestion list should only be updated when a user is not navigating the list through keying.
+		// Note: it's intentional to use object reference comparison here.
+		// Since `verticals` props is connected from a redux state here, if the two references are identical,
+		// we can safely say that the two content are identical, thanks to the immutability invariant of redux.
+		if ( prevProps.verticals !== this.props.verticals && ! this.state.isSuggestionSelected ) {
+			// It's safe here to call setState() because we prevent the indefinite loop by the wrapping condition.
+			// See the official doc here: https://reactjs.org/docs/react-component.html#componentdidupdate
+			// eslint-disable-next-line react/no-did-update-set-state
+			this.setSearchResults( this.props.verticals );
+		}
+	}
+
+	/***
+	 * Sets `state.results` with incoming HTTP results, retaining previous non-user vertical search results
+	 * if the incoming HTTP results contain only user-defined results.
+	 *
+	 * Then we'll debounce the update on the vertical while the user is typing to reduce
+	 * the amount of times we update the site preview.
+	 *
+	 * This function could better be performed in the backend eventually.
+	 *
+	 * @param {Array} results Incoming HTTP results
+	 */
+	setSearchResults = results => {
+		if ( size( results ) ) {
+			const { candidateVerticals, isSuggestionSelected } = this.state;
+			// if the only result is a user input (non-vertical),
+			// and we have some previous results
+			// then concat that with the previous results and remove the last user input
+			if (
+				! find( results, item => ! item.isUserInputVertical ) &&
+				1 < size( candidateVerticals )
+			) {
+				results = candidateVerticals.filter( item => ! item.isUserInputVertical ).concat( results );
+			}
+
+			const matches = this.searchForVerticalMatches( this.state.searchValue );
+
+			this.setState( { candidateVerticals: results }, () =>
+				// if a user has selected a vertical from the suggestion list,
+				// set a flag to load sans debounce
+				isSuggestionSelected
+					? this.updateVerticalData( matches, this.state.searchValue )
+					: this.updateVerticalDataDebounced( matches, this.state.searchValue )
```

Comment:
> why we would need to debounce state updating here

Ooh, that was a bad rebase. You're right @southp . It doesn't need to be there. Nice 👀 !!

But to avoid updating the state too frequently we need it `[onChange](https://github.com/Automattic/wp-calypso/pull/31708/files/d64a4faf7499774bb41d009b7c077a7c61c11526#diff-9553ff7100a8a69d6011c3bc6f8dded2R177)` at least.

**With no debounce:**
![no-debounce-onchange](https://user-images.githubusercontent.com/6458278/55372464-8ee92e80-554d-11e9-8bc8-f0953aa1783a.gif)

**With:**
![debounce-on-change](https://user-images.githubusercontent.com/6458278/55372504-b0e2b100-554d-11e9-8ac9-5ea8a8d88fab.gif)

I'll look more closely at this today.